Caching of file-set hashes by local path and mtimes

tclose opened the pull request:
## Types of changes

- New feature (non-breaking change which adds functionality)

## Summary

Will address #683 by

- [x] `bytes_repr()` overloads for FileSets to yield a "time-stamp" object consisting of a key (file path) and modification time as the first item in the generator.
- [x] `hash_single()` checks for these key/mtime pairs in a global "local hashes cache" dict
    - returns cached hash, if present
    - otherwise it proceeds through the remaining byte chunks, and calculates the hash
- [x] calculated hashes are saved into/loaded from the cache directory using a hash of the key/mtime 


## Checklist
<!--- Please, let us know if you need help-->
- [x] I have added tests to cover my changes (if necessary)
- [x] I have updated documentation (if necessary)

## Notes
I'm pretty happy with how this turned out with the exception of a few of wrinkles. Any suggestions would be most welcome
- There isn't a clean way to specify the location of the persistent hash cache in the top-level code or put it in the `cache_location` path (my original plan) given that `checksum` and `hash` are object properties instead of methods
    - I have therefore just dumped it in `~/.pydra-hash-cache`
    - not ideal to have to (effectively) hard-code this and having it user dependent
    - Having it in the user directory does mean that if the same cache directory is accessed from different machines (e.g. shared network drive) then local paths won't clash (although chance of mtimes being the same would be vanishingly small)
- The keys of the caches themselves are hashed and stored as files in the persistent cache directory, and will therefore never be cleaned up.
    - This could be ok as they are very small files so if it grows as time goes on it probably won't have much impact
    - This is how pydra caches work in general
- The resolution of mtime differs depending on OS, with Ubuntu and Windows having quite low resolution, i.e. order of seconds
    - I have had to put a sleep call inside the hash test to ensure the mtimes are different
    - Could cause issues if you were updating the contents of files within a loop and spinning off different workflows (not sure whether this would ever happen in practice)
    - No easy way to disable the mtime caching behaviour if this does become a problem

ghisvail commented:
I am not confident I understand enough of the previous and proposed caching methods to have an opinion and assess this PR.

AFAIK, the previous caching mechanism generated cache folders for each task (including the workflow) in a common cache location which was set to a temporary folder, unless overridden with the `cache_dir` argument.

Each task's cache folder was composed of its name and a hash value, the latter being computed from the task's input field values. The cache would be re-used, if the name and hash values did not change, and the same cache folder was specified as `cache_dir`. Otherwise, it would recompute everything in a new cache location somewhere in temp.

Could you confirm whether my summary is accurate and whether this PR is proposing to change the fundamentals of this mechanism?

tclose commented:
> I am not confident I understand enough of the previous and proposed caching methods to have an opinion and assess this PR.

No worries, I just put you all down as reviewers, so you were notified. Don't feel like you need to contribute if the area isn't familiar

> AFAIK, the previous caching mechanism generated cache folders for each task (including the workflow) in a common cache location which was set to a temporary folder, unless overridden with the `cache_dir` argument.
> 
> Each task's cache folder was composed of its name and a hash value, the latter being computed from the task's input field values. The cache would be re-used, if the name and hash values did not change, and the same cache folder was specified as `cache_dir`. Otherwise, it would recompute everything in a new cache location somewhere in temp.
> 
> Could you confirm whether my summary is accurate and whether this PR is proposing to change the fundamentals of this mechanism?

Yes, your understanding is correct. Tis is how the hashing works, both currently and in this PR. This PR seeks to improve (or more accurately, restore) performance by caching the hashes of file/directory types themselves, so files/directories don't need to be rehashed (which can be an expensive operation) each time the checksum is accessed.

It does this by hashing the path and mtime of the file/directory (not to be confused with the hashing of the file/directory contents), and using it as a key to look up a "hash cache" (not to be confused with the execution cache) that contains previously computed file/directory hashes.

effigies commented:
> * I have had to put a sleep call inside the hash test to ensure the mtimes are different

Better to mock the call than to sleep. Here's an example where I've done that with `time.time()`:
https://github.com/nipy/nibabel/blob/5f37398a2f8211c175b798eb42298b963c693ae0/nibabel/tests/test_openers.py#L457-L464

tclose commented:
> > * I have had to put a sleep call inside the hash test to ensure the mtimes are different
> 
> Better to mock the call than to sleep. Here's an example where I've done that with `time.time()`: https://github.com/nipy/nibabel/blob/5f37398a2f8211c175b798eb42298b963c693ae0/nibabel/tests/test_openers.py#L457-L464

Interesting, I assumed that the value for the mtime was controlled by the file-system not Python.

tclose commented on pydra/utils/hash.py:
```diff
@@ -51,19 +53,109 @@
+                f"Persistent cache location '{location}' is not a directory"
+            )
+
+    def get_hash(self, key: CacheKey, calculate_hash: ty.Callable) -> Hash:
```
ok, sounds good

ghisvail commented on pydra/utils/hash.py:
```diff
@@ -51,19 +53,109 @@
+def hash_function(obj, persistent_cache: ty.Optional[Path] = None):
     """Generate hash of object."""
-    return hash_object(obj).hex()
+    if persistent_cache is None:
+        # FIXME: Ideally the default location would be inside one of the cache_locations
+        # but can't think of a clean way to pass the cache_locations down to this part
+        # of the code, so just dumping in the home directory instead
+        persistent_cache = (Path("~") / ".pydra-hash-cache").expanduser()
+        try:
+            if not persistent_cache.exists():
+                persistent_cache.mkdir()
+        except Exception:
+            persistent_cache = None
```
I believe we should be careful where to place stuff in the home directory.

We could follow the old "wild west" convention of placing files in a single `.pydra` folder at the root of `$HOME`, but this is generally considered sloppy, and sooner or later someone will complain and ask whether we can support established standards for their platform of choice.

Luckily, there is a [package](https://pypi.org/project/platformdirs/) implementing this already. I guess we would be concerned by this [section](https://platformdirs.readthedocs.io/en/latest/api.html#cache-directory).

djarecka replied:
- perhaps we could just use a specific environmental variable if specified by the user?
- Since this is meant to be used within specific run, should we have option to clean the files that are e.g. a week old?

tclose replied:
That package looks good @ghisvail, although I not sure this is worth adding another dependency for, what are everyone's thoughts?

In either case, an env var is a good idea @djarecka. Do we have a naming convention for them?

> Since this is meant to be used within specific run, should we have option to clean the files that are e.g. a week old?

I don't see this needing to be constrained to a specific run, but probably no harm to run a clean-up of week/month-old files after a workflow completes

ghisvail replied:
> although I not sure this is worth adding another dependency for

Every piece of software which cache stuff to home end up writing (usually badly) some subset of what's covered in `platformdirs`. It's not worth adding if we are confident to implement it *right*. If not, the package is dependency free and [well tested](https://github.com/platformdirs/platformdirs/actions/workflows/check.yml), so the risks and additional bloat to Pydra would be minimal.

> we could just use a specific environmental variable if specified by the user

We could indeed (say `PYDRA_CACHE_DIR`), but the suitable fallback value is platform dependent. That's where the value of `platformdirs` comes in.

> I don't see this needing to be constrained to a specific run, but probably no harm to run a clean-up of week/month-old files after a workflow completes

I am trying to picture a use case for this: if I run workflow A and months later workflow B, should I expect Pydra to clean cached files up from workflow A (and any other which are old enough w.r.t. workflow B)?

tclose replied:
> > although I not sure this is worth adding another dependency for
> 
> Every piece of software which cache stuff to home end up writing (usually badly) some subset of what's covered in `platformdirs`. It's not worth adding if we are confident to implement it _right_. If not, the package is dependency free and [well tested](https://github.com/platformdirs/platformdirs/actions/workflows/check.yml), so the risks and additional bloat to Pydra would be minimal.

Sounds ok to me, I looked into pulling out just the bits we need from that package, and for Mac+Linux it was trivial but Windows is pretty complicated so decided against it.

> > we could just use a specific environmental variable if specified by the user
> 
> We could indeed (say `PYDRA_CACHE_DIR`), but the suitable fallback value is platform dependent. That's where the value of `platformdirs` comes in.

👍

> 
> > I don't see this needing to be constrained to a specific run, but probably no harm to run a clean-up of week/month-old files after a workflow completes
> 
> I am trying to picture a use case for this: if I run workflow A and months later workflow B, should I expect Pydra to clean cached files up from workflow A (and any other which are old enough w.r.t. workflow B)?

Yes, if they are not touched by Workflow B and they are over a month old they get cleaned up.

tclose replied:
It is still showing up as "1 change requested" from you, I think it might have been the method name change (which I implemented in the last round of changes). Is that right?

tclose replied:
@ghisvail I realised that *platformdirs* is already dependency of *black* and *virtualenv* so have used it to locate the hash cache as you suggested.

ghisvail reviewed (changes requested):
Some suggestions and a question mark regarding the cache path in the home folder. Maybe also worth having a look to mtime mocking instead of sleep steps as suggested by Chris.

Looks solid otherwise 👍

djarecka commented:
@tclose - you can try with new slurm testing workflow, should work now!

djarecka reviewed (approved):
lgtm! just left comments regarding the location of the cache directory

tclose commented on pydra/utils/hash.py:
```diff
@@ -51,19 +53,109 @@
+        return Hash(hsh)
+
+
+def persistent_cache_converter(
+    path: ty.Union[Path, str, PersistentCache, None]
+) -> PersistentCache:
```
I believe this is a strange limitation of attrs, that converters can't be classmethods. I presume it is in order to preserve performance

tclose replied:
Actually, I had this wrong. It is just a limitation of mypy. I have added it as a classmethod and just ignored the mypy warning

tclose commented:
> > * I have had to put a sleep call inside the hash test to ensure the mtimes are different
> 
> Better to mock the call than to sleep. Here's an example where I've done that with `time.time()`: https://github.com/nipy/nibabel/blob/5f37398a2f8211c175b798eb42298b963c693ae0/nibabel/tests/test_openers.py#L457-L464

@effigies I'm not sure this works on Windows (it seems to work on Ubuntu)

tclose commented:
@effigies, I think that I have addressed the outstanding issues with this PR so it just needs your review. However, I had to revert the mtime mocking as it doesn't appear to work for Windows (see https://github.com/nipype/pydra/actions/runs/6307685788/job/17124695852), unless you have some ideas on how to do it.

djarecka commented:
@tclose - the recent commit to this PR comes from some rebasing?

tclose commented:
> @tclose - the recent commit to this PR comes from some rebasing?

I rebased it on top of master after the environment PR was merged and fixed up a few things

djarecka commented:
ok, trying to figure out if I should review it again, since I already accepted

effigies commented on pydra/utils/hash.py:
```diff
@@ -52,19 +54,144 @@
+        return PersistentCache(path)
+
+
+@attrs.define
+class Cache:
+    persistent: ty.Optional[PersistentCache] = attrs.field(
+        default=None,
+        converter=PersistentCache.from_path,  # type: ignore[misc]
+    )
+    _hashes: ty.Dict[int, Hash] = attrs.field(factory=dict)
+
+    def __getitem__(self, object_id: int) -> Hash:
+        return self._hashes[object_id]
+
+    def __setitem__(self, object_id: int, hsh: Hash):
+        self._hashes[object_id] = hsh
+
+    def __contains__(self, object_id):
+        return object_id in self._hashes
```
`self.persistent` never gets used. How is this supposed to work?

tclose replied:
It is used on L241 (https://github.com/tclose/pydra/blob/a031ea5dfc1571d2171e0df4157bb97114b9b1eb/pydra/utils/hash.py#L241)

effigies reviewed (commented):
Sorry, I've completely lost the plot on this PR. I don't understand what's going on, but there are some things that look wrong to me.

tclose commented on pydra/utils/hash.py:
```diff
@@ -52,19 +54,144 @@
+            return self._hashes[key]
+        except KeyError:
+            pass
+        key_path = self.location / blake2s(str(key).encode()).hexdigest()
```
I think I just copied it from somewhere. I don't really know the pros & cons of different methods. Which would you recommend?

djarecka replied:
looks like we were using `blake2s` before, but indeed doesn't seem to be the best choice (probably copied from somewhere). We can just use `blake2 ` or try [blake3](https://github.com/BLAKE3-team/BLAKE3), but could be also done in a separate pr

djarecka commented:
sorry, I'm sure we discussed it at some point, but I'm not sure about one important thing...  looks like if I move file around my filesystem I have no way of reusing the previous tasks now, is that right?

djarecka commented:
sorry, I was wrong, the task has correct hash when the file is the same with a different path. 

Perhaps you can add this test: https://github.com/djarecka/pydra/blob/db782c20890797bd58eb1d52545b7104f7d41aa4/pydra/engine/tests/test_node_task.py#L1568

(I was planning to create a PR to you, but I must have merged to the branch more things to my branch)

tclose commented:
> sorry, I was wrong, the task has correct hash when the file is the same with a different path.
> 
> Perhaps you can add this test: https://github.com/djarecka/pydra/blob/db782c20890797bd58eb1d52545b7104f7d41aa4/pydra/engine/tests/test_node_task.py#L1568
> 
> (I was planning to create a PR to you, but I must have merged to the branch more things to my branch)

Ok, nice addition

djarecka commented:
I've just realized that we should also have a similar test when the persistent cache is used, but realized that it's not enough to set `PYDRA_HASH_CACHE`. Sorry, for asking for more work, but can we have some test when the persistent cache is used together with running task or workflow

tclose commented:
@djarecka I have added a new test to ensure that the persistent cache gets hit during the running of tasks. Let me know if there is anything else you need me to do

djarecka commented on pydra/engine/tests/test_node_task.py:
```diff
@@ -1599,3 +1604,62 @@ def test_task_files_cachelocations(plugin_dask_opt, tmp_path):
+        return super().contents
+
+
+def test_task_files_persistentcache(tmp_path):
```
@tclose - where are you setting the persistent cach path? i.e. `PYDRA_HASH_CACHE`

tclose replied:
At @ghisvail's suggestion, the hash cache is stored in a system-dependent user cache directory using `platformdirs.user_cache_dir` by default (e.g. `/Users/<username>/Library/Caches/pydra/<version-number>` on MacOS).

I have just tweaked the code so that it is now put in a `hashes` subdirectory of that user cache dir (accessible in the `pydra.utils.user_cache_dir` variable) just in case any other cache data needs to be stored at some point in the future.

djarecka replied:
oh, got it! Sorry I missed that. I've just realized that I made a typo when I was setting `PYDRA_HASH_CACHE` and that's why it was not saving the hashes there, and was confused where this is being saved..

djarecka commented:
@tclose - thanks so much for the work!
